client/src/hooks.ts

```diff
@@ -171,3 +171,25 @@ export function usePing() {
     }
   }, [isOnline, user]);
 }
+
+function getIsDocumentHidden() {
+  if (typeof document !== "undefined") {
+    return !document.hidden;
+  }
+  return false;
+}
+
+export function usePageVisibility() {
+  const [isVisible, setIsVisible] = React.useState(getIsDocumentHidden());
+  const onVisibilityChange = () => setIsVisible(getIsDocumentHidden());
+  useEffect(() => {
+    if (typeof document !== "undefined") {
+      const visibilityChange = "visibilitychange";
+      document.addEventListener(visibilityChange, onVisibilityChange, false);
+      return () => {
+        document.removeEventListener(visibilityChange, onVisibilityChange);
+      };
+    }
+  });
+  return isVisible;
+}
```

Comment:
This whole section of code is a bit weird, how about:
```suggestion
export function usePageVisibility() {
  const [isVisible, setIsVisible] = React.useState(
    typeof document === "undefined"
      ? false
      : document.visibilityState === "visible"
  );
  useEffect(() => {
    const onVisibilityChange = () =>
      setIsVisible(document.visibilityState === "visible");
    document.addEventListener("visibilitychange", onVisibilityChange);
    return () => {
      document.removeEventListener("visibilitychange", onVisibilityChange);
    };
  }, []);
  return isVisible;
}
}
```